parser/internal/bytestring/bytestring.go

```diff
@@ -194,3 +199,42 @@ func (s *String) ReadUint64(out *uint64) bool {
 		uint64(v[4])<<32 | uint64(v[5])<<40 | uint64(v[6])<<48 | uint64(v[7])<<56
 	return true
 }
+
+// ReadScriptInt64 reads and interprets a Bitcoin-custom compact integer
+// encoding used for int64 numbers in scripts.
+//
+// See https://github.com/zcash/zcash/blob/4df60f4b334dd9aee5df3a481aee63f40b52654b/src/script/script.h#L363-L378
+func (s *String) ReadScriptInt64(num *int64) bool {
+	// First byte is either an integer opcode, or the number of bytes in the
+	// number.
+	firstBytes := s.read(1)
+	if firstBytes == nil {
+		return false
+	}
+	firstByte := firstBytes[0]
+
+	var number uint64
+
+	if firstByte == OP_1NEGATE {
+		*num = -1
+		return true
+	} else if firstByte == OP_0 {
+		number = 0
+	} else if firstByte >= OP_1 && firstByte <= OP_16 {
```

Comment:
It's not clear that this is correct. The C++ code seems to be directly comparing to the numbers 1 and 16, not these opcode values.

Reply:
That's intentional. The referenced C++ code implements the serialization side; this function implements the parsing side. Parsing in `zcashd` would happen in the bowels of the script system.

Reply:
I'll update the comments to clarify this, and reference a partial parser.